cloudbridge/cloud/providers/gce/resources.py

```diff
@@ -1063,3 +1068,253 @@ def delete(self):
                                          address=self._ip['name'])
                                  .execute())
        self._provider.wait_for_operation(response, region=self._region)
+
+
+class GCEVolume(BaseVolume):
+
+    VOLUME_STATE_MAP = {
+        'RESTORING': VolumeState.CONFIGURING,
+        'PENDING': VolumeState.CONFIGURING,
+        'READY': VolumeState.AVAILABLE,
+        'DONE': VolumeState.AVAILABLE,
+        'RUNNING': VolumeState.IN_USE,
+    }
+
+    def __init__(self, provider, volume):
+        super(GCEVolume, self).__init__(provider)
+        self._volume = volume
+
+    @property
+    def id(self):
+        return self._volume.get('selfLink')
+
+    @property
+    def name(self):
+        """
+        Get the volume name.
+        """
+        return self._volume.get('name')
+
+    @name.setter
+    # pylint:disable=arguments-differ
+    def name(self, value):
+        """
+        Set the volume name.
+        """
+        raise NotImplementedError('Not supported by this provider.')
+
+    @property
+    def description(self):
+        return self._volume.get('description')
+
+    @description.setter
+    def description(self, value):
+        raise NotImplementedError('Not supported by this provider.')
+
+    @property
+    def size(self):
+        return self._volume.get('sizeGb')
+
+    @property
+    def create_time(self):
+        return self._volume.get('creationTimestamp')
+
+    @property
+    def zone_id(self):
+        return self._volume.get('zone')
+
+    @property
+    def source(self):
+        return self._volume.get('sourceSnapshot')
+
+    @property
+    def attachments(self):
+        if 'users' in self._volume and len(self._volume) > 0:
+            return BaseAttachmentInfo(self,
+                                      self._volume.get('users')[0],
+                                      None)
+        else:
+            return None
+
+    def attach(self, instance, device):
+        """
+        Attach this volume to an instance.
+
+        Parameter `device` is ignored here. The user need to mount the disk
+        so that the operating system can use the available storage space.
+        https://cloud.google.com/compute/docs/disks/add-persistent-disk
+        """
+        attach_disk_body = {
+            "source": self.id
+        }
+        instance_name = instance.name if isinstance(
+            instance,
+            GCEInstance) else instance
+        response = (self._provider.gce_compute
+                        .instances()
+                        .attachDisk(
+                            project=self._provider.project_name,
+                            zone=self._provider.default_zone,
+                            instance=instance_name,
+                            body=attach_disk_body).execute())
+
+    def detach(self, force=False):
+        """
+        Detach this volume from an instance.
+        """
+        if not self.attachments:
```

Comment:
Please add some commenting here about this first block where you are checking multiple ways whether the volume is actually attached to an instance. It isn't immediately obvious why all the checks are needed.